test/unit_test_runner.d

```diff
@@ -0,0 +1,392 @@
+#!/usr/bin/env rdmd
+module unit_test_runner;
+
+import std.algorithm : filter, map, joiner, substitute;
+import std.array : array, join;
+import std.conv : to;
+import std.exception : enforce;
+import std.file : dirEntries, exists, SpanMode, mkdirRecurse, write;
+import std.format : format;
+import std.getopt : getopt;
+import std.path : buildNormalizedPath, buildPath, dirName, dirSeparator,
+    stripExtension, setExtension;
+import std.process : environment, spawnProcess, spawnShell, wait;
+import std.range : empty;
+import std.stdio;
+import std.string : join, outdent;
+
+version (Posix)
+    enum exeExtension = "";
+else version (Windows)
+    enum exeExtension = ".exe";
+
+version(Windows)
+    enum os ="windows";
+else version(OSX)
+    enum os ="osx";
+else version(linux)
+    enum os ="linux";
+else version(FreeBSD)
+    enum os ="freebsd";
+else version(OpenBSD)
+    enum os ="openbsd";
+else version(NetBSD)
+    enum os ="netbsd";
+else version(DragonFlyBSD)
+    enum os ="dragonflybsd";
+else version(Solaris)
+    enum os ="solaris";
+else version(SunOS)
+    enum os ="solaris";
+else
+    static assert(0, "Unrecognized or unsupported OS.");
+
+enum scriptDir = __FILE_FULL_PATH__.dirName.buildNormalizedPath;
+enum projectRoot = scriptDir.buildNormalizedPath("..");
+enum unitTestDir = scriptDir.buildNormalizedPath("unit");
+enum resultsDir = testPath("test_results");
+enum generatedDir = projectRoot.buildPath("generated");
```

Comment:
A lot of duplication happened here. Why not put common parts into `tools`?

Reply:
What is `tools`? Is that something that already exists?